feat: Add provide* combinators to Http and HttpResult

frekw opened the pull request:
This follows up on the work done in #189 and instead adds the `provide*` combinators to the `HttpResult`.

I also added all of the combinators to `Http` itself, just acting as wrappers on top.

frekw committed:
feat: Add provide* combinators to HttpResult

frekw committed:
fix: Formatting

frekw committed:
fix: scalafix

frekw committed:
fix: Scala3

frekw committed:
feat: Add all provide* combinators to http

frekw committed:
fix: Style

frekw committed:
fix: Fix warning

frekw committed:
fix: Remove unnecessary comment

tusharmath commented on zio-http/src/main/scala/zhttp/http/Http.scala:
```diff
@@ -204,6 +266,50 @@ object Http {
   )                                                                             extends Http[R, EE, A, BB]
 
+  private final case class Provide[R, E, A, B](self: Http[R, E, A, B], r: R) extends Http[Any, E, A, B] {
+    override private[zhttp] def execute(a: A): HttpResult[Any, E, B] =
```
`execute` was supposed to be final (miss from my side).
You will loose out on stacksafety if you override it.

frekw replied:
Gotcha.

The only reason why I use it is to get around an (possibly incorrect) exhaustivity warning on Scala3, see d45475f. Would be happy to just inline all impls if someone can point me to another way around this!

frekw replied:
1a119f3 is the only way around this I've found that also makes the compiler happy. Can't say it's super pretty (and not sure if extracting `doProvide` keeps it stack-safe?). If you have any other ideas of how to address this I'd be very grateful 🙏

frekw replied:
Oh. And that doesn't seem to work for 2.13. :/

frekw replied:
Changing all of them like so:

```scala
+      case p @ ProvideLayer(_, _)     => p.doProvide
+      // case p: ProvideLayer[_, _, _, _, _, _] => p.doProvide
```

works for both versions but emits warnings in Scala 3. 🤷

frekw committed:
fix: Don't override execute/evaluate

tusharmath commented on zio-http/src/main/scala/zhttp/http/HttpResult.scala:
```diff
@@ -94,11 +201,13 @@ object HttpResult {
-      case Suspend(r)              => evaluate(r())
-      case FoldM(self, ee, aa, dd) =>
+      case m: Out[_, _, _]                       => m
+      case Suspend(r)                            => evaluate(r())
```
Similar to `Suspend` you can add a case for `Provide(a, r)`, it should look something like —

```scala
  case Provide(a, r) => 
    a match {
      case a: Out[_, _, _] => a // For `Out` type `provide` is not required.
      case Effect(z)       => evaluate(z.provide(r)) // For `ZIO` use the built-in `provide` operator
      case FoldM(self, ee, aa, dd) => self.foldM(e => ee(e).provide(r), a => (aa).provide(r), dd.provide(r)) // For `foldM` handle env for all three channels — Failure + Success + Default.
    }
```

frekw committed:
fix: Update based on #265

frekw committed:
fix: Formatting

frekw committed:
fix: Scala 2.12

frekw commented:
@tusharmath I updated this to be based on #265. I've removed `provideSomeLayer` and `provideCustomLayer` because they were giving me too much of a headache due to missing evidence for `R <: Has[_]` in `evaluate`. I also think these combinators are quite valuable in themselves.

However, as with the code in #265 we're still seeing exhaustivity warnings in Scala3 (that I believe are false) but am unsure how to work around them. Will ask around a bit on the Scala gitter.

frekw commented:
Could potentially be https://github.com/lampepfl/dotty/issues/9682 that we're running into.

frekw committed:
fix: Extract Provide/ProvideLayer evaluate

frekw committed:
fix: Make scala3 happy

tusharmath commented:
fixed in #364
